xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,721 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.Listener;
+import io.envoyproxy.envoy.api.v2.RouteConfiguration;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.api.v2.route.Route;
+import io.envoyproxy.envoy.api.v2.route.VirtualHost;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager;
+import io.envoyproxy.envoy.config.filter.network.http_connection_manager.v2.Rds;
+import io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.alts.GoogleDefaultChannelBuilder;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.Bootstrapper.ChannelCreds;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class XdsClientImpl extends XdsClient {
+  private static final Logger logger = Logger.getLogger(XdsClientImpl.class.getName());
+
+  @VisibleForTesting
+  static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.api.v2.Listener";
+  @VisibleForTesting
+  static final String ADS_TYPE_URL_RDS =
+      "type.googleapis.com/envoy.api.v2.RouteConfiguration";
+
+  private final ManagedChannel channel;
+  private final SynchronizationContext syncContext;
+  private final ScheduledExecutorService timeService;
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final Stopwatch stopwatch;
+  // The node identifier to be included in xDS requests. Management server only requires the
+  // first request to carry the node identifier on a stream. It should be identical if present
+  // more than once.
+  private final Node node;
+
+  // Each Listener received in LDS responses indicates an RouteConfiguration. Since we are
+  // always requesting for a singleton Listener resource, we only need to remember the
+  // a single Listener information.
+  // Upon receiving an LDS response, the entry for the RouteConfiguration indicated by the
```

Comment:
suggestion: How about moving these detailed comments to where this logic is implemented instead of next to member declaration?

Reply:
+1. There are so many special restrictions on so many fields, it's very hard to read.

Reply:
ACK.